README.md
SGG_SKSpineImport
=================

This is a library to import json animations created in Esoteric Software's Spine application.

DISCLAIMER:

I have no affiliation with esoteric Software whatsoever and built this runtime without the use of their supplied starting point. I am still relatively new to this whole *programming* thing and therefore may not have implemented everything in the most efficient or optimized manner. Therefore, I will not be held responsible for problems pertaining to this runtime in any way. Also, if something could be done in a better way, that's the reason I open sourced this. Feel free to make changes and make a pull request.

And finally, in all sincerity, I hope that this project either helps you or makes your life a little easier. :)

Usage:

Examples and shortcomings can be viewed by building the project. It's configured to demo for OS X, but builds for iOS and can be reconjiggered to use iOS controls.

Import the entire SpineImporter folder to your project. Add the following line to the header of whatever you need to add the runtime to:

	#import "SpineImport.h"
	
Create spine objects as such:

	boy = [SGG_Spine node];
	[boy skeletonFromFileNamed:@"spineboy" andAtlasNamed:@"spineboy" andUseSkinNamed:Nil];
	boy.position = CGPointMake(self.size.width/4, self.size.height/4);
	boy.queuedAnimation = @"walk";
	boy.queueIntro = 0.1;
	[boy runAnimation:@"walk" andCount:0 withIntroPeriodOf:0.1 andUseQueue:YES];
	boy.zPosition = 0;
	[self addChild:boy];

And then the animations must be "activated" for each spine object in the update method of the scene:

	[boy activateAnimations];
	
More examples are contained in the project. I believe it should be fairly straightforward.

To import image assets, you must add the ORIGINAL image files to your project in an Xcode atlas file. Atlases created by Spine or third parties are unsupported.

Features:

*   Import Spine animations
*	Handles fairly easily
*	Support for multiple skins on a skeleton (click to see the alt goblin skin, using the same skeleton)
	*	Also supports partial replacement (e.g. just hands)
*	Start and stop animation
*	Slot Animation
*	Z Ordering (but not animation)
*	Queue system for animation (when you jump, it will automatically return to running/walking when the animation finishes)
*	Animations can be transitioned into each other (as opposed to hard cutting from one to the other)
*	**Supports non linear keyframe timing modes**
*	**Supports differing playback speeds**

*	Builds for OS X and iOS


Limitations:

*	~~Slot animation gets confused when you switch skins during an animation (in the example build, hold the mouse down for a while, then let go and watch the goblin eyes)~~
*	~~Animation is built using SKAction - therefore, easing of keyframes is limited to the simple enumerated versions of easy ease that SKAction includes (ease in, ease out, ease both, linear)~~
	*	~~This runtime automatically detects the closest approximation and applies it, however~~ See above
*	Z order keyframes are NOT implemented
*	Mixing animation (have your feet walking while your arms swing a weapon, for example) is not implemented.
*	Requires using the built in atlas management.
*	Notice how the elf stays down and doesn't reset his animation. I'm not sure why that is. I'm suspecting that the animation file itself is wonky, but I haven't delved into it.
*	Can't have slashes (/) in image names
*	I'm not sure if the animation sequence vs queuing system order of operations is the most efficient or logical or if another configuration would be superior.
	*	Currently, the priority is 
		1.	runAnimation method (called directly)
		2.	running the next animation in a sequence
		3.	running whatever animation is queued for repeat



Requirements:

*	Xcode 5+
*	SpriteKit project

Target requirements:

*	iOS 7+
or 
*	Mac OS X 10.9+ or whatever SpriteKit's requirements are. I'm not 100% sure and I don't care enough to look them up just for this. I'm not your mom.


## Method and Property Highlights

#### debugMode *property*
When set to YES, prints debug information in NSLog and shows a frame counter on the animation

#### isRunningAnimation *property*
Readonly BOOL that can tell you if an animation is currently running.

#### currentAnimationSequence *property*
If there is a sequence of animations running, this is an NSArray that will list their names. Will be nil if there is no sequence running. Does not activate for queued animations.

#### currentAnimation *property*
When an animation is running, this will be a read only NSString that will tell you the name of the current animation.

#### animationCount *property*
If an animation is set to repeat X number of times (before resorting to the queue or stopping), this will return the amount of times remaining.

#### queuedAnimation *property*
Use this NSString to set the queued animation to run after the current animation finishes. (Example, set this to "walk" to always walk, but transition to "jump" anytime a button is pressed).

#### queueIntro *property*
CGFloat value in seconds (affected by playbackSpeed) to transition from the previous animation to the queued animation. Will be ignored if the queued animation is the same as the previous animation that completed.

#### useQueue *property*
BOOL value that can be toggled to turn animation queuing on and off.

#### timeResolution *property*
CGFloat value to determine the "framerate" at which animations should be calculated. Default value is 1/120. This doesn't affect the playback speed.

For example, at 1/120 and a framerate of 1/60, the playback will skip approximately every other frame. The reason it's at 1/120 is to create smoother playback in the event of unevenly timed rendering of frames. It could very likely be lowered to 1/60 and see no harm.

If you intend to use slower motion, you will need more granularity and should set it to a "higher" denominator value.

#### playbackSpeed *property*
CGFloat value to determine how fast to play back the animation. *Can* be ramped, but you may notice stuttering.

#### currentFrame *property*
Read only NSInteger value to tell you what the currentFrame of animation being displayed is.


#### -skeletonFromFileNamed:andAtlasNamed:andUseSkinNamed:
Since the object will already have been created with a line such as

	spineSkeleton = [SGG_Spine node];

this method is used to set up the attachments, animations, bones, and all the other cool stuff Spine lets you do. This should be done AFTER the following property(ies) are set:

*	timeResolution

-(void)skeletonFromFileNamed:(NSString\*)name andAtlasNamed:(NSString\*)atlasName andUseSkinNamed:(NSString*)skinName;

###### Parameters
*	*name*
	*	NSString naming the file the skeleton resides in
*	*atlasName*
	*	NSString naming the .atlas file to pull imagery from
*	*skinName*
	*	NSString naming the skin to use. Defaults to "default" if nil.



#### -runAnimation:andCount:withIntroPeriodOf:andUseQueue:
Sets the spine object to play back the animation with the name you provide.

-(void)runAnimation:(NSString*)animationName andCount:(NSInteger)count withIntroPeriodOf:(const CGFloat)introPeriod andUseQueue:(BOOL)useQueue;

###### Parameters
*	*animationName*
	*	NSString value of animation to play back
*	*count*
	*	NSInteger value of number of times to repeat the animation (0 plays once, 1 plays twice)
*	*introPeriod*
	*	CGFloat value in seconds of duration to transition from the current status to the start of the new animation
*	*useQueue*
	*	BOOL value to determine whether to return to the queued animation after this animation finishes


#### -runAnimation:andCount:

-(void)runAnimation:(NSString*)animationName andCount:(NSInteger)count;

This is a shorthand of the previous method. Uses current setting for "useQueue".

###### Parameters
*	*animationName*
	*	NSString value of animation to play back
*	*count*
	*	NSInteger value of number of times to repeat the animation (0 plays once, 1 plays twice)
	
#### -runAnimationSequence:andUseQueue:
-(void)runAnimationSequence:(NSArray *)animationNames andUseQueue:(BOOL)useQueue;

Creates a sequence of animations with no transitions between them.

###### Parameters
*	*animationNames*
	*	NSArray of NSStrings naming the animations in order of playback
*	*useQueue*
	*	BOOL value to determine whether to return to the queued animation after this animation finishes

#### -stopAnimation

-(void)stopAnimation;

Stops the current animation (if there is one).

#### -jumpToFrame:
-(void)jumpToFrame:(NSInteger)frame;

Stops animation and jumps to a frame in the animation.

###### Parameters
*	*frame*
	*	NSInteger value of frame to jump to. Remember to take into account the timeResolution.
	
#### -jumpToNextFrame

Stops animation (if running) and jumps to the next frame. Remember to take into account the timeResolution.

#### -jumpToPreviousFrame

Stops animation (if running) and jumps to the previous frame. Remember to take into account the timeResolution.

#### -activateAnimations

Required method to playback animations. Must run every frame of the app (at least when animations are supposed to run), so it's best to put it in the scene update method.

#### -resetSkeleton
Resets the spine object back to its default state.

#### -changeSkinTo
-(void)changeSkinTo:(NSString*)skin;

Changes the current skin of the entire spine object to a new skin.

###### Parameters
*	*skin*
	*	NSString value name of the skin you want to change to
	

#### -changeSkinPartial:
-(void)changeSkinPartial:(NSDictionary *)slotsToReplace;

Replaces the skin of any nodes mentioned in the slotsToReplace dictionary with their accompanying skins.

###### Parameters
*	*slotsToReplace*
	*	NSDictionary with keys as the names of the slots you want to change and the value being the name of the skin to change to. Example:

```
	NSDictionary* partReplacement = @{@"part": @"skinName", @"torso": @"goblin", @"head": @"goblin"};
	[goblin changeSkinPartial:partReplacement];
```

#### -resetSkinPartial
-(void)resetSkinPartial;

Resets any changes made through changeSkinPartial; reverts the skeleton back to the skins used when the scene was created.
Should only be called after changeSkinPartial.


#### -changeTexturePartial:
-(void)changeTexturePartial:(NSDictionary *)attachmentsToReplace;

Replaces the texture of any nodes mentioned in the attachmentsToReplace dictionary with their accompanying texture.

###### Parameters
*	*attachmentsToReplace*
	*	NSDictionary with keys as the names of the attachments you want to change and the value being the name of the attachments to change to. Example:

```
	NSDictionary* partReplacement = @{@"attachmentToReplace": @"attachementToReplaceWith", @"torso": @"goblin-torso", @"head": @"goblin-head"};
	[goblin changeTexturePartial:partReplacement];
```

#### -resetTexturePartial
-(void)resetTexturePartial;

Resets any changes made through resetTexturePartial; reverts the skeleton back to the textures used when the scene was created.
Should only be called after changeTexturePartial.



#### -colorizeSlots:withColor:andIntensity:
-(void)colorizeSlots:(NSArray \*)slotsToColorize withColor:(SKColor \*)color andIntensity:(CGFloat)blendFactor;

Colorizes the slots listed in the slotsToColorize array using the specified color and blend factor.

Can be used to make a body part flash red, or to let the player change their hair/skin color.

###### Parameters
*	*slotsToColorize*
	*	NSArray of NSStrings naming the slots to colorize
*	*color*
	*	SKColor of the color to colorize with.
*	*blendFactor*
	*	CGFloat value from 0.0 to 1.0 to determine how much to colorize the slots.

#### -resetColorizedSlots
-(void)resetColorizedSlots;

Resets any changes made through resetColorizedSlots; reverts the skeleton back to the colors used when the scene was created.
Should only be called after colorizeSlots.


#### -findBoneNamed:
-(SGG_SpineBone\*)findBoneNamed:(NSString\*)boneName;

Finds and returns an SGG_SpineBone object (inherits from SKNode) with the name provided.

###### Parameters
*	*boneName*
	*	NSString value of the name of the bone you are looking for

## Questions or comments?
Let me know!

##### Additional Resource
https://github.com/massivepenguin/SGG_SKSpineImport


changelog.txt

Version 1.1

	Features
	- added non linear key frame support
	- modfied changeSkinPartial method and renamed to changeTexturePartial (if you rely on the old method, use changeTexturePartial as the syntax should be the same)
	- replaced changeSkinPartial with new method (changes the skin setting for each individiual slot instead of specific textures, which SHOULD be superior)
	- removed queueCount property as it seemed superfluous with being able to sequence animations now
	- added frame stepping forward and backward
	- added playback speed control
	- added ability to colorize all slots at once
	- probably other stuff I forgot to list here


	Bug Fixes
	- because of the new animation engine, slots are properly animated when changing skins
	
	Known Issues
	-

Version 1.0

	- Initial Release

SpineTesting/SGG_MyScene.m
//
//  SGG_MyScene.m
//  SpineTesting
//
//  Created by Michael Redig on 2/10/14.
//  Copyright (c) 2014 Secret Game Group LLC. All rights reserved.
//

#import "SGG_MyScene.h"
#import "SpineImport.h"
#import "SGG_SpineBoneAction.h"

@interface SGG_MyScene () {
	SGG_Spine* boy;
	SGG_Spine* elf;
	SGG_Spine* goblin;
	SGG_Spine* stepTest;
	
	CGPoint startLocation;
	CGPoint ballPosition;
	
	BOOL enabled;
	
}

@end

@implementation SGG_MyScene {

}

-(id)initWithSize:(CGSize)size {	
	if (self = [super initWithSize:size]) {
		/* Setup your scene here */
		
		
		boy = [SGG_Spine node];
//		boy.debugMode = YES;
//		boy.timeResolution = 1.0 / 1200.0; // this is typically overkill, 1/120 will normally be MORE than enough, but this demo can go to some VERY slow motion. 1/120 is also the default.
		[boy skeletonFromFileNamed:@"spineboy" andAtlasNamed:@"spineboy" andUseSkinNamed:Nil];
		boy.position = CGPointMake(self.size.width/4, self.size.height/4);
//		[boy runAnimationSequence:@[@"walk", @"jump", @"walk", @"walk", @"jump"] andUseQueue:NO]; //uncomment to see how a sequence works (commment the other animation calls)
		boy.queuedAnimation = @"walk";
		boy.name = @"boy";
		boy.queueIntro = 0.1;
		[boy runAnimation:@"walk" andCount:0 withIntroPeriodOf:0.1 andUseQueue:YES];
		boy.zPosition = 0;
		[self addChild:boy];

		
		
		elf = [SGG_Spine node];
		[elf skeletonFromFileNamed:@"elf" andAtlasNamed:@"elf" andUseSkinNamed:Nil];
		elf.position = CGPointMake(self.size.width/2, self.size.height/4);
		[elf runAnimation:@"standing" andCount:-1];
		elf.queuedAnimation = @"standing";
		elf.queueIntro = 0.1;
		elf.zPosition = 20;
		elf.xScale = 0.6;
		elf.yScale = 0.6;
		[self addChild:elf];

		
		goblin = [SGG_Spine node];
//		goblin.debugMode = YES;
		[goblin skeletonFromFileNamed:@"goblins" andAtlasNamed:@"goblin" andUseSkinNamed:@"goblin"];
		goblin.position = CGPointMake((self.size.width/4)*3, self.size.height/4);
		[goblin runAnimation:@"walk" andCount:-1];
		goblin.zPosition = 10;
		[self addChild:goblin];
		

		



		SKLabelNode* label1 = [SKLabelNode labelNodeWithFontNamed:@"Helvetica Neue Light"];
		label1.text = @"space to make boy jump";
		label1.color = [SKColor whiteColor];
		label1.position = CGPointMake(self.size.width/2, self.size.height/4 - 20);
		[self addChild:label1];
		
		SKLabelNode* label2 = [SKLabelNode labelNodeWithFontNamed:@"Helvetica Neue Light"];
		label2.text = @"\"a\" and \"d\" to change directions for boy";
		label2.color = [SKColor whiteColor];
		label2.position = CGPointMake(self.size.width/2, label1.position.y - 30);
		[self addChild:label2];
		
		SKLabelNode* label3 = [SKLabelNode labelNodeWithFontNamed:@"Helvetica Neue Light"];
		label3.text = @"click to change goblin skin";
		label3.color = [SKColor whiteColor];
		label3.position = CGPointMake(self.size.width/2, label2.position.y - 30);
		[self addChild:label3];
		
		SKLabelNode* label4 = [SKLabelNode labelNodeWithFontNamed:@"Helvetica Neue Light"];
		label4.text = @"\"j\", \"k\", and \"l\" to change playback speed for boy";
		label4.color = [SKColor whiteColor];
		label4.position = CGPointMake(self.size.width/2, label3.position.y - 30);
		[self addChild:label4];
		
	}
	return self;
}


#if TARGET_OS_IPHONE

-(void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
	CGPoint location = [[touches anyObject] locationInNode:self];
	[self inputBegan:location];
}

-(void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
	CGPoint location = [[touches anyObject] locationInNode:self];
	[self inputMoved:location];
}

-(void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
	CGPoint location = [[touches anyObject] locationInNode:self];
	[self inputEnded:location];
}


#else

-(void)mouseDown:(NSEvent *)theEvent {
	 /* Called when a mouse click occurs */
	
	CGPoint location = [theEvent locationInNode:self];
	[self inputBegan:location];

}

-(void)mouseDragged:(NSEvent *)theEvent {
	CGPoint location = [theEvent locationInNode:self];
	[self inputMoved:location];

}

-(void)mouseUp:(NSEvent *)theEvent {
	CGPoint location = [theEvent locationInNode:self];
	[self inputEnded:location];

}

-(void)keyDown:(NSEvent *)theEvent {
	
	NSString *characters = [theEvent characters];
	if ([characters length]) {
		for (int s = 0; s<[characters length]; s++) {
			unichar character = [characters characterAtIndex:s];
			switch (character) {
				case ' ':{
					if (![boy.currentAnimation isEqualToString:@"jump"]) {
						[boy runAnimation:@"jump" andCount:0 withIntroPeriodOf:0.1 andUseQueue:YES];
					}
					
				}
					break;
				case 'a': boy.xScale = -1;
					break;
				case 'd':boy.xScale = 1;
					break;
				case 's':[boy stopAnimation];
					break;
				case 'w':[boy runAnimation:@"walk" andCount:0 withIntroPeriodOf:0.1 andUseQueue:YES];
					break;
				case 'l':boy.playbackSpeed += 0.05;
					break;
				case 'j':boy.playbackSpeed -= 0.05;
					break;
				case 'k':boy.playbackSpeed = 1;
					break;
			}
		}
	}
	
	
	
}

#endif


-(void)inputBegan:(CGPoint)location {

	startLocation = location;
	
	[elf runAnimation:@"trip" andCount:0 withIntroPeriodOf:0.1 andUseQueue:YES];
	

	
	NSArray* partsToColorize = @[@"head", @"left shoulder", @"torso"];
	SKColor* color = [SKColor redColor];

	[goblin changeSkinTo:@"goblingirl"];
	[boy colorizeSlots:partsToColorize withColor:color andIntensity:1];
	
	[goblin colorizeAllSlotsWithColor:[SKColor blueColor] andIntensity:0.8];
/*
//this section is a correct example of partial texture replacement
	NSDictionary* partReplacement = @{@"torso": @"goblin-torso", @"head": @"goblin-head"};

	[goblin changeTexturePartial:partReplacement];
*/
	
/*
 //this section is a correct example of partial skin replacement

	NSDictionary* skinReplacement = @{@"torso": @"goblingirl", @"head": @"goblingirl", @"eyes": @"goblingirl"};
	[goblin changeSkinPartial:skinReplacement];
	
*/



	
}

-(void)inputMoved:(CGPoint)location {

	if (location.y > startLocation.y + 20) {
		if (![boy.currentAnimationSequence[0] isEqualToString:@"jump"]) {
			[boy runAnimation:@"jump" andCount:0 withIntroPeriodOf:0.1 andUseQueue:YES];
		}
	}
	
	if (location.x > startLocation.x + 20) {
		boy.xScale = 1;
	} else if (location.x < startLocation.x -20) {
		boy.xScale = -1;
	}

	
}

-(void)inputEnded:(CGPoint)location {

	[goblin changeSkinTo:@"goblin"];
	
//	[goblin resetSkinPartial];
	[goblin resetTexturePartial];
	
	//reset the colors
	[boy resetColorizedSlots];
	[goblin resetColorizedSlots];
	
	//[goblin resetSkeleton];
	
}



-(void)update:(CFTimeInterval)currentTime {
	/* Called before each frame is rendered */
	
	//each spine object must be "activated" here to playback their animation
	[goblin activateAnimations];
	[boy activateAnimations];
	[elf activateAnimations];

	
}



//NSValue testing
/*
 CGFloat xa = arc4random() % 100;
 CGFloat xb = arc4random() % 100;
 CGFloat ya = arc4random() % 100;
 CGFloat yb = arc4random() % 100;
 
 xb = xb / 100;
 yb = yb / 100;
 
 CGFloat x = xa + xb;
 CGFloat y = ya + yb;
 
 bool xc = arc4random() % 2;
 bool yc = arc4random() % 2;
 
 if (xc) {
 x *= -1;
 }
 if (yc) {
 y *= -1;
 }
 
 [[[SGG_SpineBoneAction alloc] init] addTranslationAtTime:0 withPoint:CGPointMake(x, y) andCurveInfo:@[@"1.2, 2.1"]];*/



@end
